File: python/taflow/lower_low.py
"""Persistent lower-low relation."""

from typing import Any

from ._bar_relation_adapter import BarRelationAdapter
from ._native import LowerLowOperator


class LowerLow(BarRelationAdapter):
    """Return 1 when a bar's low is below the preceding bar's low.

    ``high`` and ``low`` are required aligned series. The first output is
    ``NaN`` because no preceding bar exists. All calculations run in Rust.
    """

    _native_cls = LowerLowOperator

    def append(self, high: float, low: float) -> "LowerLow":
        super().append(high, low)
        return self

    def extend(self, high: Any, low: Any) -> "LowerLow":
        super().extend(high, low)
        return self

    def reset(self) -> "LowerLow":
        super().reset()
        return self

File: crates/taflow-core/src/stream/lower_low.rs
//! Persistent lower-low relation.

use super::bar_relation::BarRelation;

/// Emit `1` when the current low is below the previous low, otherwise `0`.
#[derive(Debug, Clone, Default)]
pub struct LowerLow {
    relation: BarRelation,
}

impl LowerLow {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn append(&mut self, high: f64, low: f64) -> Option<f64> {
        self.relation
            .append(high, low, |_, l, _, previous_l| l < previous_l)
    }

    pub fn value(&self) -> Option<f64> {
        self.relation.value()
    }

    pub fn reset(&mut self) {
        self.relation.reset();
    }
}
